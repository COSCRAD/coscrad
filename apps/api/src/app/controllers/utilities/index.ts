// TODO Barrel export everything in this directory and fix existing imports
export * from './mixLinkIntoViewModelDescription';
