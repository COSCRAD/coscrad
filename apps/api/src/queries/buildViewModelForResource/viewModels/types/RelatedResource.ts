import { EdgeConnectionMember } from '../../../../domain/models/context/edge-connection.entity';

export type RelatedResource = EdgeConnectionMember;
