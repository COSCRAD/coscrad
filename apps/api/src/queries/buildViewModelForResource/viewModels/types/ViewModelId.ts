export type ViewModelId = string;

export type HasViewModelId = {
    id: ViewModelId;
};
