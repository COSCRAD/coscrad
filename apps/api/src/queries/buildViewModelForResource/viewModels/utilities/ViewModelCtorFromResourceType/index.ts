export * from './aggregateTypeToViewModelCtor';
export * from './types';
