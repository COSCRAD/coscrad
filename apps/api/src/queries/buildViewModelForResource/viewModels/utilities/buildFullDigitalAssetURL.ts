export default (baseURL: string, filename: string, extension = 'mp3'): string => {
    return `${baseURL}${filename}.${extension}`;
};
