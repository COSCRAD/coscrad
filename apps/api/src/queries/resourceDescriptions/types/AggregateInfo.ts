import { AggregateType, IAggregateInfo } from '@coscrad/api-interfaces';

export type AggregateInfo = IAggregateInfo<AggregateType>;
