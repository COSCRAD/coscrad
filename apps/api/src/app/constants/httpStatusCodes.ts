import { ValueType } from '../../lib/types/valueType';

const httpStatusCodes = {
    ok: 200,
    badRequest: 400,
    unauthorized: 401,
    forbidden: 403,
    notFound: 404,
    internalError: 500,
};

export default httpStatusCodes;

export type HttpStatusCode = ValueType<typeof httpStatusCodes>;
