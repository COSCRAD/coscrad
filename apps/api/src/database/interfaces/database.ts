import { EntityId } from '../../domain/types/entity-id';
import { Maybe } from '../../lib/types/maybe';

export interface IDatabase {
  fetchById: <TEntityDTO>(
    id: EntityId,
    collectionName: string
  ) => Promise<Maybe<TEntityDTO>>;

  fetchMany: <TEntityDTO>(
    collectionName: string
  ) => Promise<Maybe<TEntityDTO[]>>;

  create: <TEntityDTO>(
    dto: TEntityDTO,
    collectionName: string
  ) => Promise<void>;

  createMany: <TEntityDTO>(
    dto: TEntityDTO[],
    collectionName: string
  ) => Promise<void>;

  getCount: (collectionName: string) => Promise<number>;

  update: <TEntityDTO>(
    id: EntityId,
    dto: TEntityDTO,
    collectionName: string
  ) => Promise<void>;
}
