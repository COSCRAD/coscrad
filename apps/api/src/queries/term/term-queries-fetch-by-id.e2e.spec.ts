import {
    AggregateType,
    CoscradUserRole,
    IDetailQueryResult,
    ITermViewModel,
    LanguageCode,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import httpStatusCodes from '../../app/constants/httpStatusCodes';
import setUpIntegrationTest from '../../app/controllers/__tests__/setUpIntegrationTest';
import getValidAggregateInstanceForTest from '../../domain/__tests__/utilities/getValidAggregateInstanceForTest';
import buildDummyUuid from '../../domain/models/__tests__/utilities/buildDummyUuid';
import { AccessControlList } from '../../domain/models/shared/access-control/access-control-list.entity';
import { TermCreated } from '../../domain/models/term/commands';
import {
    ITermQueryRepository,
    TERM_QUERY_REPOSITORY_TOKEN,
} from '../../domain/models/term/queries';
import { CoscradContributor } from '../../domain/models/user-management/contributor';
import { CoscradUserGroup } from '../../domain/models/user-management/group/entities/coscrad-user-group.entity';
import { CoscradUserWithGroups } from '../../domain/models/user-management/user/entities/user/coscrad-user-with-groups';
import { CoscradUser } from '../../domain/models/user-management/user/entities/user/coscrad-user.entity';
import { FullName } from '../../domain/models/user-management/user/entities/user/full-name.entity';
import { AggregateId } from '../../domain/types/AggregateId';
import { clonePlainObjectWithOverrides } from '../../lib/utilities/clonePlainObjectWithOverrides';
import { ArangoDatabaseProvider } from '../../persistence/database/database.provider';
import TestRepositoryProvider from '../../persistence/repositories/__tests__/TestRepositoryProvider';
import generateDatabaseNameForTestSuite from '../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import buildTestDataInFlatFormat from '../../test-data/buildTestDataInFlatFormat';
import { TestEventStream } from '../../test-data/events';
import { BaseResourceViewModel } from '../buildViewModelForResource/viewModels/base-resource.view-model';
import { TermViewModel } from '../buildViewModelForResource/viewModels/term.view-model';

// Set up endpoints: index endpoint, id endpoint
const indexEndpoint = `/resources/terms`;

const buildDetailEndpoint = (id: AggregateId) => `${indexEndpoint}/${id}`;

// We require an existing user for ACL tests
const dummyQueryUserId = buildDummyUuid(4);

const { user: users, userGroup: userGroups } = buildTestDataInFlatFormat();

const dummyUser = (users as CoscradUser[])[0].clone({
    authProviderUserId: `auth0|${dummyQueryUserId}`,
    id: dummyQueryUserId,
    roles: [CoscradUserRole.viewer],
});

const dummyGroup = (userGroups as CoscradUserGroup[])[0].clone({ userIds: [dummyUser.id] });

const dummyUserWithGroups = new CoscradUserWithGroups(dummyUser, [dummyGroup]);

// Set up test data (use event sourcing to set up state)
const termText = `Term (in the language)`;

const originalLanguage = LanguageCode.Haida;

const termId = buildDummyUuid(1);

const termCompositeId = {
    type: AggregateType.term,
    id: termId,
};

const dummyContributorFirstName = 'Dumb';

const dummyContributorLastName = 'McContributor';

const dummyContributor = getValidAggregateInstanceForTest(AggregateType.contributor).clone({
    fullName: new FullName({
        firstName: dummyContributorFirstName,
        lastName: dummyContributorLastName,
    }),
});

const termCreated = new TestEventStream().andThen<TermCreated>({
    type: 'TERM_CREATED',
    payload: {
        text: termText,
        languageCode: originalLanguage,
    },
    meta: {
        contributorIds: [dummyContributor.id],
    },
});

const dummyTerm = TermViewModel.fromTermCreated(termCreated.as(termCompositeId)[0] as TermCreated);

const targetTermView = clonePlainObjectWithOverrides(dummyTerm, {
    isPublished: true,
    contributions: [
        {
            id: dummyContributor.id,
            fullName: dummyContributor.fullName.toString(),
        },
    ],
});

const privateTermThatUserCanAccess = clonePlainObjectWithOverrides(dummyTerm, {
    accessControlList: new AccessControlList({
        allowedUserIds: [dummyQueryUserId],
        allowedGroupIds: [],
    }),
    isPublished: false,
});

// TODO Add happy path cases for a prompt term
// const promptTermId = buildDummyUuid(2)

const assertResourceHasContributionFor = (
    { id: contributorId }: CoscradContributor,
    resource: BaseResourceViewModel
) => {
    const hasContribution = resource.contributions.some(
        ({ id }) => id === contributorId
        // TODO support joining in contributors
        // && fullName === `${firstName} ${lastName}`
    );

    expect(hasContribution).toBe(true);
};

describe(`when querying for a term: fetch by Id`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    let termQueryRepository: ITermQueryRepository;

    let seedTerms: (terms: TermViewModel[]) => Promise<void>;

    // let eventPublisher: ICoscradEventPublisher;
    beforeEach(async () => {
        await testRepositoryProvider.testSetup();

        await databaseProvider.clearViews();
    });

    afterAll(async () => {
        await app.close();

        databaseProvider.close();
    });

    describe(`when the user is unauthenticated`, () => {
        beforeAll(async () => {
            ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                {
                    ARANGO_DB_NAME: testDatabaseName,
                }
                // no authenticated user
            ));

            // eventPublisher = app.get(EVENT_PUBLISHER_TOKEN);

            termQueryRepository = app.get(TERM_QUERY_REPOSITORY_TOKEN);

            /**
             * We need to use the proper publisher to make sure events only
             * run if they match the pattern for the given handler.
             */
            seedTerms = async (terms: TermViewModel[]) => {
                await termQueryRepository.createMany(terms);
            };
        });

        describe(`when there is a term with the given Id`, () => {
            describe(`when a term is published`, () => {
                beforeEach(async () => {
                    /**
                     * It is important that we do this before allowing the event
                     * handlers to run, as they will only add contributions
                     * if the corresponding contributor exists.
                     */
                    await testRepositoryProvider
                        .getContributorRepository()
                        .create(dummyContributor);

                    await databaseProvider.clearViews();

                    await seedTerms([targetTermView]);
                });

                it('should return the expected result', async () => {
                    const res = await request(app.getHttpServer()).get(buildDetailEndpoint(termId));

                    expect(res.status).toBe(httpStatusCodes.ok);

                    const result = res.body as IDetailQueryResult<TermViewModel>;

                    expect(result.id).toBe(termId);

                    // We don't expose actions to non-admin users
                    expect(result.actions).toEqual([]);

                    assertResourceHasContributionFor(dummyContributor, result);
                });
            });

            describe(`when a term is not published`, () => {
                beforeEach(async () => {
                    // note that there is no publication event in this event history

                    // TODO: we need to check that contributors come through
                    await seedTerms([
                        clonePlainObjectWithOverrides(targetTermView, {
                            isPublished: false,
                        }),
                    ]);
                });

                // We pretend the resource does not exist when the user
                // does not have access to this term
                it(`should return not found (404)`, async () => {
                    const res = await request(app.getHttpServer()).get(buildDetailEndpoint(termId));

                    expect(res.status).toBe(httpStatusCodes.notFound);
                });
            });
        });

        describe(`when there is no term with the given Id`, () => {
            it(`should return not found (404)`, async () => {
                const res = await request(app.getHttpServer()).get(
                    buildDetailEndpoint(buildDummyUuid(456))
                );

                expect(res.status).toBe(httpStatusCodes.notFound);
            });
        });
    });

    describe(`when the user is authenticated`, () => {
        describe(`when the user is a coscrad admin`, () => {
            beforeAll(async () => {
                ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                    {
                        ARANGO_DB_NAME: testDatabaseName,
                    },
                    {
                        testUserWithGroups: new CoscradUserWithGroups(
                            dummyUser.clone({
                                roles: [CoscradUserRole.superAdmin],
                            }),
                            []
                        ),
                    }
                ));

                seedTerms = async (terms: TermViewModel[]) => {
                    await termQueryRepository.createMany(terms);
                };
            });

            describe(`when there is a term with the given Id`, () => {
                describe(`when the term is published`, () => {
                    beforeEach(async () => {
                        // TODO: we need to check that contributors come through
                        await seedTerms([targetTermView]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        // Commands should be visible to admin
                        expect(res.body.actions).not.toEqual([]);
                    });
                });

                describe(`when the term is not published`, () => {
                    beforeEach(async () => {
                        // note that there is no publication event in this event history

                        // TODO: we need to check that contributors come through

                        await seedTerms([
                            clonePlainObjectWithOverrides(targetTermView, {
                                isPublished: false,
                            }),
                        ]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        // Commands should be visible to admin
                        expect(res.body.actions).not.toEqual([]);
                    });
                });

                describe(`when the term is not published but the user has explicit access`, () => {
                    beforeEach(async () => {
                        // note that there is no publication event in this event history

                        // TODO: we need to check that contributors come through

                        await seedTerms([privateTermThatUserCanAccess]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        // Commands should be visible to admin
                        expect(res.body.actions).not.toEqual([]);
                    });
                });
            });

            describe(`when there is no term with the given Id`, () => {
                it(`should return not found (404)`, async () => {
                    const res = await request(app.getHttpServer()).get(
                        buildDetailEndpoint(buildDummyUuid(456))
                    );

                    expect(res.status).toBe(httpStatusCodes.notFound);
                });
            });
        });

        describe(`when the user is a project admin`, () => {
            beforeAll(async () => {
                ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                    {
                        ARANGO_DB_NAME: testDatabaseName,
                    },
                    {
                        testUserWithGroups: new CoscradUserWithGroups(
                            dummyUser.clone({
                                roles: [CoscradUserRole.projectAdmin],
                            }),
                            []
                        ),
                    }
                ));

                seedTerms = async (terms: TermViewModel[]) => {
                    await termQueryRepository.createMany(terms);
                };
            });

            describe(`when there is a term with the given Id`, () => {
                describe(`when the term is published`, () => {
                    beforeEach(async () => {
                        // TODO: we need to check that contributors come through

                        await seedTerms([targetTermView]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        const term = res.body as IDetailQueryResult<ITermViewModel>;

                        const { actions } = term;

                        // admin should see commands
                        expect(actions).not.toHaveLength(0);

                        /**
                         * TODO We should add a separate test that checks
                         * that the correct actions come through in different
                         * scenarios.
                         */
                        expect(actions).toMatchSnapshot();
                    });
                });

                describe(`when the term is not published`, () => {
                    beforeEach(async () => {
                        await seedTerms([
                            clonePlainObjectWithOverrides(targetTermView, {
                                isPublished: false,
                            }),
                        ]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        // Commands should be visible to admin
                        expect(res.body.actions).not.toEqual([]);
                    });
                });

                describe(`when the term is not published but the user has explicit access`, () => {
                    // This case is a bit unclear: does this project admin have access to
                    // the project this term is a part of?
                    beforeEach(async () => {
                        await seedTerms([privateTermThatUserCanAccess]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        // Commands should be visible to admin
                        expect(res.body.actions).not.toEqual([]);
                    });
                });
            });

            describe(`when there is no term with the given Id`, () => {
                it(`should return not found (404)`, async () => {
                    const res = await request(app.getHttpServer()).get(
                        buildDetailEndpoint(buildDummyUuid(456))
                    );

                    expect(res.status).toBe(httpStatusCodes.notFound);
                });
            });
        });

        describe(`when the user is an ordinary authenticated user`, () => {
            beforeAll(async () => {
                ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
                    {
                        ARANGO_DB_NAME: testDatabaseName,
                    },
                    {
                        testUserWithGroups: dummyUserWithGroups,
                    }
                ));

                seedTerms = async (terms: TermViewModel[]) => {
                    await termQueryRepository.createMany(terms);
                };
            });

            describe(`when there is a term with the given Id`, () => {
                describe(`when the term is published`, () => {
                    beforeEach(async () => {
                        await seedTerms([targetTermView]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        // We don't expose actions to non-admin users
                        expect(res.body.actions).toEqual([]);
                    });
                });

                describe(`when the term is not published and the user does not have access`, () => {
                    beforeEach(async () => {
                        await seedTerms([
                            clonePlainObjectWithOverrides(targetTermView, {
                                isPublished: false,
                                // no special access here
                                accessControlList: new AccessControlList(),
                            }),
                        ]);
                    });

                    // We pretend the resource does not exist when the user
                    // does not have access to this term
                    it(`should return not found (404)`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.notFound);
                    });
                });

                describe(`when the term is not published but the user has access`, () => {
                    beforeEach(async () => {
                        await seedTerms([privateTermThatUserCanAccess]);
                    });

                    it(`should return the expected result`, async () => {
                        const res = await request(app.getHttpServer()).get(
                            buildDetailEndpoint(termId)
                        );

                        expect(res.status).toBe(httpStatusCodes.ok);

                        // We don't expose actions to non-admin users
                        expect(res.body.actions).toEqual([]);
                    });
                });
            });

            describe(`when there is no term with the given Id`, () => {
                it(`should return not found (404)`, async () => {
                    const res = await request(app.getHttpServer()).get(
                        buildDetailEndpoint(buildDummyUuid(456))
                    );

                    expect(res.status).toBe(httpStatusCodes.notFound);
                });
            });
        });
    });
});
