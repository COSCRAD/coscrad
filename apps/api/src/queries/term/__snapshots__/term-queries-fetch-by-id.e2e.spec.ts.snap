// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a term: fetch by Id when the user is authenticated when the user is a project admin when there is a term with the given Id when the term is published should return the expected result 1`] = `
[
  {
    "description": "Make a resource visible to the public",
    "form": {
      "fields": [],
      "prepopulatedFields": {
        "aggregateCompositeIdentifier": {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
          "type": "term",
        },
      },
    },
    "label": "Publish Resource",
    "type": "PUBLISH_RESOURCE",
  },
  {
    "description": "creates a note about this particular resource",
    "form": {
      "fields": [
        {
          "constraints": [],
          "description": "system-wide unique identifier for the resource about which we are making a note",
          "label": "CompositeIdentifier",
          "name": "resourceCompositeIdentifier",
          "type": "JSON_INPUT",
        },
        {},
        {
          "constraints": [
            {
              "message": "must be a non-empty string",
              "name": "non-empty string",
            },
            {
              "message": "must be a defined value",
              "name": "defined value",
            },
          ],
          "description": "text for the note",
          "label": "text",
          "name": "text",
          "type": "TEXT_FIELD",
        },
        {
          "constraints": [
            {
              "message": "Must be a valid \${propertyLabel}",
              "name": "IS_ENUM",
            },
            {
              "message": "Required",
              "name": "defined value",
            },
          ],
          "description": "the language in which you are writing the note",
          "label": "language code",
          "name": "languageCode",
          "options": [
            {
              "display": "Chilcotin",
              "value": "clc",
            },
            {
              "display": "Haida",
              "value": "hai",
            },
            {
              "display": "English",
              "value": "en",
            },
            {
              "display": "French",
              "value": "fra",
            },
            {
              "display": "Chinook",
              "value": "chn",
            },
            {
              "display": "Zapotec",
              "value": "zap",
            },
            {
              "display": "Spanish",
              "value": "spa",
            },
          ],
          "type": "STATIC_SELECT",
        },
      ],
      "prepopulatedFields": {
        "aggregateCompositeIdentifier": {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
          "type": "term",
        },
      },
    },
    "label": "Create Note",
    "type": "CREATE_NOTE_ABOUT_RESOURCE",
  },
  {
    "description": "connects two resources with a note and context",
    "form": {
      "fields": [
        {
          "constraints": [],
          "description": "system-wide unique identifier for the resource to which you are making a connection",
          "label": "to member composite identifier",
          "name": "toMemberCompositeIdentifier",
          "type": "JSON_INPUT",
        },
        {},
        {
          "constraints": [],
          "description": "system-wide unique identifier for the resource from which you are making a connection",
          "label": "from member composite identifier",
          "name": "fromMemberCompositeIdentifier",
          "type": "JSON_INPUT",
        },
        {},
        {
          "constraints": [
            {
              "message": "must be a non-empty string",
              "name": "non-empty string",
            },
            {
              "message": "must be a defined value",
              "name": "defined value",
            },
          ],
          "description": "text for the note",
          "label": "text",
          "name": "text",
          "type": "TEXT_FIELD",
        },
        {
          "constraints": [
            {
              "message": "Must be a valid \${propertyLabel}",
              "name": "IS_ENUM",
            },
            {
              "message": "Required",
              "name": "defined value",
            },
          ],
          "description": "the language in which you are writing the note",
          "label": "language code",
          "name": "languageCode",
          "options": [
            {
              "display": "Chilcotin",
              "value": "clc",
            },
            {
              "display": "Haida",
              "value": "hai",
            },
            {
              "display": "English",
              "value": "en",
            },
            {
              "display": "French",
              "value": "fra",
            },
            {
              "display": "Chinook",
              "value": "chn",
            },
            {
              "display": "Zapotec",
              "value": "zap",
            },
            {
              "display": "Spanish",
              "value": "spa",
            },
          ],
          "type": "STATIC_SELECT",
        },
      ],
      "prepopulatedFields": {
        "aggregateCompositeIdentifier": {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
          "type": "term",
        },
      },
    },
    "label": "Connect Resources with Note",
    "type": "CONNECT_RESOURCES_WITH_NOTE",
  },
  {
    "description": "Translate an existing term (typically from the language to English)",
    "form": {
      "fields": [
        {
          "constraints": [
            {
              "message": "must be a non-empty string",
              "name": "non-empty string",
            },
            {
              "message": "must be a defined value",
              "name": "defined value",
            },
          ],
          "description": "translation for the given term",
          "label": "translation",
          "name": "translation",
          "type": "TEXT_FIELD",
        },
        {
          "constraints": [
            {
              "message": "Must be a valid \${propertyLabel}",
              "name": "IS_ENUM",
            },
            {
              "message": "Required",
              "name": "defined value",
            },
          ],
          "description": "language in which you are translating the term (typically English)",
          "label": "language code",
          "name": "languageCode",
          "options": [
            {
              "display": "Chilcotin",
              "value": "clc",
            },
            {
              "display": "Haida",
              "value": "hai",
            },
            {
              "display": "English",
              "value": "en",
            },
            {
              "display": "French",
              "value": "fra",
            },
            {
              "display": "Chinook",
              "value": "chn",
            },
            {
              "display": "Zapotec",
              "value": "zap",
            },
            {
              "display": "Spanish",
              "value": "spa",
            },
          ],
          "type": "STATIC_SELECT",
        },
      ],
      "prepopulatedFields": {
        "aggregateCompositeIdentifier": {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
          "type": "term",
        },
      },
    },
    "label": "Translate Term",
    "type": "TRANSLATE_TERM",
  },
  {
    "description": "add audio for an existing term",
    "form": {
      "fields": [
        {
          "constraints": [
            {
              "message": "must be a UUID",
              "name": "UUID",
            },
            {
              "message": "must be a defined value",
              "name": "defined value",
            },
          ],
          "description": "reference to an audio item",
          "label": "audio item ID",
          "name": "audioItemId",
          "options": {
            "aggregateType": "audioItem",
          },
          "type": "DYNAMIC_SELECT",
        },
        {
          "constraints": [
            {
              "message": "Must be a valid \${propertyLabel}",
              "name": "IS_ENUM",
            },
            {
              "message": "Required",
              "name": "defined value",
            },
          ],
          "description": "the language of the audio",
          "label": "language code",
          "name": "languageCode",
          "options": [
            {
              "display": "Chilcotin",
              "value": "clc",
            },
            {
              "display": "Haida",
              "value": "hai",
            },
            {
              "display": "English",
              "value": "en",
            },
            {
              "display": "French",
              "value": "fra",
            },
            {
              "display": "Chinook",
              "value": "chn",
            },
            {
              "display": "Zapotec",
              "value": "zap",
            },
            {
              "display": "Spanish",
              "value": "spa",
            },
          ],
          "type": "STATIC_SELECT",
        },
      ],
      "prepopulatedFields": {
        "aggregateCompositeIdentifier": {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
          "type": "term",
        },
      },
    },
    "label": "add audio for term",
    "type": "ADD_AUDIO_FOR_TERM",
  },
]
`;
